### PR TITLE
Support for cross-platform C# Grpc for .Net Bigtable driver

### PR DESCRIPTION
**Pull request by cj525**
The project would not build x64, and the Grpc.Core nuget hard-linked a 32 bit implementation.  I create a client for Bigtable using grpc and protobuf and need it to work cross-platform/architecture.  I have made changes to that end, most notably IPlatformInvocation, although some casting work in the core C projects was needed as well.  Software CLA has been signed.

### CODE REVIEW

**Comment by cj525** [comment]
"I signed it!"  (however I signed under my work account, which is linked in github, but your bot may not realize that)  PM for details, pls

**Comment by jtattermusch** [comment]
Hi, thanks for the pull request! It is great to see such engagement.

Please note that I am currently actively working on x64 support as well and some of the changes you've suggested are also part of my existing pull requests (see #4512).
I am also trying to make x64 builds on VS2013 and VS2015 run continuously on Jenkins, so that we can be sure it works in the future.
Recently, we've also changed the definition of gpr_timespec type, which should result in much simpler P/Invoking (#4397).

I'll try to go through the PR and put some comments. Unfortunately, after today, I'll be OOO until Jan 12, so please don't expect much progress on this PR until then.

Btw, the pull request needs conflict resolution.

File: src/core/security/credentials.c [hunk 1]
```diff
@@ -180,7 +180,7 @@ void grpc_server_credentials_set_auth_metadata_processor(
       "grpc_server_credentials_set_auth_metadata_processor("
       "creds=%p, "
       "processor=grpc_auth_metadata_processor { process: %lx, state: %p })",
-      3, (creds, (unsigned long)processor.process, processor.state));
```
Comment (jtattermusch) [inline src/core/security/credentials.c]: Should be taken care of by my previous PRs (same for other C core chagnes).

File: src/csharp/Grpc.Core/GrpcEnvironment.cs [hunk 2]
```diff
@@ -45,16 +46,10 @@ namespace Grpc.Core
     /// </summary>
     public class GrpcEnvironment
     {
-        const int THREAD_POOL_SIZE = 4;
+        const int DefaultThreadPoolSize = 4;
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/GrpcEnvironment.cs]: Please send name and style polishing like this in a separate pull request.

File: src/csharp/Grpc.Core/GrpcEnvironment.cs [hunk 3]
```diff
@@ -67,6 +62,21 @@ public class GrpcEnvironment
...
 
+
+        /// <summary>
+        /// This is provided for advanced use-cases, such as libraries which provide a layer of abstract over grpc.
+        /// Setting this value will not have an effect until after all open channels/servers are shutdown.
+        /// Therefore, an application may change this at anytime but will need to orchastrate restrarting all channels/servers.
+        /// </summary>
+        /// <remarks>The default is 4</remarks>
+        /// <returns>The previous thread pool size</returns>
+        public static int SetThreadPoolSize(int numberOfThreads)
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/GrpcEnvironment.cs]: I can't accept such change until we do more work on performance testing (as the internals of the thread pool used might change and then we would need to break the API).

Also, this is a logically separate change, so there should be a separated PR for it (but as I said, I can't probably accept this yet).

File: src/csharp/Grpc.Core/Internal/AsyncCallBase.cs [hunk 4]
```diff
@@ -157,8 +157,9 @@ protected void StartReadMessageInternal(AsyncCompletionDelegate<TRead> completio
                 Preconditions.CheckNotNull(completionDelegate, "Completion delegate cannot be null");
                 CheckReadingAllowed();
 
-                call.StartReceiveMessage(HandleReadFinished);
+                // I feel like a prerequisite variable should be set before entering that code path
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/AsyncCallBase.cs]: why?

File: src/csharp/Grpc.Core/Internal/PlatformInvocation.cs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    /// <summary>
+    /// Platform invocation dispatch
+    /// </summary>
+    internal static class PlatformInvocation
+    {
+        /// <summary>
+        /// Retrieve an implementation of platform invocations specific to the current cpu's architecture.
+        /// This allows assemblies to function correctly when compiled for AnyCPU
+        /// </summary>
+        internal static IPlatformInvocation Implementation { get; private set; }
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/PlatformInvocation.cs]: I understand the idea, but can you please summarize what exactly are the differences in different implementations of IPlatformInvocations?  It seems like 99% of the stuff is identical. Can we focus on looking at the differences and figuring out what we can do to prevent the need to have the same definition copy and pasted 3 times?

Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/PlatformInvocation.cs]: I understand the idea, but can you please summarize what exactly are the differences in different implementations of IPlatformInvocations?  It seems like 99% of the stuff is identical. Can we focus on looking at the differences and figuring out what we can do to prevent the need to have the same definition copy and pasted 3 times?

File: src/csharp/Grpc.Core/Internal/PlatformInvocationWin64.cs [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+        void IPlatformInvocation.grpcsharp_completion_queue_destroy(IntPtr cq)
+        {
+            grpcsharp_completion_queue_destroy(cq);
+        }
+
+
+
+        // CStringSafeHandle 
+
+        [DllImport("grpc_csharp_ext-x64.dll")]
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/PlatformInvocationWin64.cs]: Ah, I see what this trick was to differentiate between 32bit and x64.
This seems like one of the possible approaches, but I was actually thinking of doing something along the path of  inspecting what the size of   IntPtr is (4 or 8) and inspecting the current platform (Linux vs Mac vs Windows) and selecting the right shared library programmatically ("dlopen" or similar).
Once the shared library is loaded, all the PInvokes should work fine and .NET won't go ahead to search for a library that was already loaded.
What do you think?

File: src/csharp/Grpc.Core/Logging/NullLogger.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Logging/NullLogger.cs]: If you with to contribute this, please do so in a separate PR.

File: vsprojects/vcxproj/grpc_csharp_ext/grpc_csharp_ext.vcxproj [hunk 8]
```diff
@@ -101,6 +101,7 @@
       <SubSystem>Windows</SubSystem>
       <GenerateDebugInformation Condition="!$(Jenkins)">true</GenerateDebugInformation>
       <GenerateDebugInformation Condition="$(Jenkins)">false</GenerateDebugInformation>
+      <OutputFile>$(OutDir)$(TargetName)-x64$(TargetExt)</OutputFile>
```
Comment (jtattermusch) [inline vsprojects/vcxproj/grpc_csharp_ext/grpc_csharp_ext.vcxproj]: please note that this file is autogenerated, you'd need to change to template under templates/ to make this work properly.

Also, a change to the .nuspec file would be necessary to make sure all the libraries are properly copied to the output folder once someone imports gRPC as a nuget packages and builds his project.

File: src/core/security/credentials.c (hunk 1 above)
Comment (cj525) [inline src/core/security/credentials.c]: I will break out the core changes I need to Bigtable.NET from x64 compilation issues.

File: src/csharp/Grpc.Core/Internal/AsyncCallBase.cs (hunk 4 above)
Comment (cj525) [inline src/csharp/Grpc.Core/Internal/AsyncCallBase.cs]: The return from native which is started by StartReceiveMessage will use readCompletionDelegate and it seemed like a race condition that this member would be set after calling StartReceiveMessage.  Albeit an unlikely race condition.  I will remove this when I separate my pull requests; it is not prerequisite functionality for what I'm trying to accomplish.

File: src/csharp/Grpc.Core/Internal/PlatformInvocation.cs (hunk 5 above)
Comment (cj525) [inline src/csharp/Grpc.Core/Internal/PlatformInvocation.cs]: The only difference is the binding for the DllImport annotation.  Without this, you can only use this library from a matching architecture and platform, which is 32-bit Windows.  After compiling with this change, I was able to compile my driver with AnyCPU and the same build worked in all use-cases (most notably Linux as that is my primary target, although I would like Bigtable.NET to just work in all cases).  Pragmas would have worked but would require different builds.  Also, with an interface the native functionality can be mocked for testing, so I opted for this approach.

File: src/csharp/Grpc.Core/Internal/PlatformInvocationWin64.cs (hunk 6 above)
Comment (cj525) [inline src/csharp/Grpc.Core/Internal/PlatformInvocationWin64.cs]: That would result in less nearly-duplicate code, and you could reduce the number of p/invoke calls this way, although because you still need to inspect the run-time environment, there will still be a need for an abstraction for the different targets.  It's not clear to me if that would be cleaner because DllImport takes care of path resolution (/lib:/usr/lib:..) on linux, etc.. and also takes care of freeing the library, which is internal code that you would need to reproduce.  I wouldn't recommend this because of the move to DNX/CoreCLR.  They seem to be in favor of DllMap, which is a monoism: https://github.com/dotnet/coreclr/issues/930, although that topic spurred a rather large thread which is still unresolved.

File: vsprojects/vcxproj/grpc_csharp_ext/grpc_csharp_ext.vcxproj (hunk 8 above)
Comment (cj525) [inline vsprojects/vcxproj/grpc_csharp_ext/grpc_csharp_ext.vcxproj]: Understood.  I was too focused on making it work and did not look more closely at this.

I will also separate the different functionality into separate pull requests.  Thank you for taking the time to look into this.

**Comment by jtattermusch** [comment]
@cj525, looks like the changes in this PR have been superseded by  https://github.com/grpc/grpc/pull/4730, which provides both 64bit and 32bit support on Windows as well a mechanism for loading the right version of the native library on linux and mac.
In the future, I plan to ship precompiled libraries for linux, mac and windows directly in the nuget package.

I am going to close this PR now, but please rest assured that contributions as well as feedback from you are still more than welcome.

File: src/csharp/Grpc.Core/Internal/AsyncCallBase.cs (hunk 4 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/AsyncCallBase.cs]: Do you mind creating a separate pull request for this?

Comment (cj525) [inline src/csharp/Grpc.Core/Internal/AsyncCallBase.cs]: Done, #4890
